packages/library-base/src/nativeDarwin/kotlin/io/realm/kotlin/internal/platform/SystemUtils.kt

```diff
@@ -56,17 +56,18 @@ public actual fun epochInSeconds(): Long =
  */
 @Suppress("MagicNumber")
 internal actual fun currentTime(): RealmInstant {
-    val secs = NSDate().timeIntervalSince1970
-    val millis = (secs * 1000 + if (secs > 0) 0.5 else -0.5).toLong()
-    return if (millis < RealmInstant.MIN.epochSeconds * 1000) {
-        RealmInstant.MIN
-    } else if (millis > RealmInstant.MAX.epochSeconds * 1000) {
-        RealmInstant.MAX
-    } else {
-        RealmInstantImpl(
-            millis.floorDiv(1000.toLong()),
-            (millis.mod(1000.toLong()) * 1000).toInt()
-        )
+    val secs: Double = NSDate().timeIntervalSince1970
+    return when {
+        // We can't convert the MIN value to ms - it is initialized with Long.MIN_VALUE and
+        // multiplying it by 1000 will cause overflow. We have to compare directly against seconds
+        secs < RealmInstant.MIN.epochSeconds -> RealmInstant.MIN
+        // Similarly here, compare to seconds instead to avoid overflow with Long.MAX_VALUE
+        secs > RealmInstant.MAX.epochSeconds -> RealmInstant.MAX
+        else -> {
+            val millis = (secs * 1000 + if (secs > 0) 0.5 else -0.5).toLong()
```

Comment:
I think that we could use the Duration framework to simplify these calculations.